src/Sort.c

```diff
@@ -144,6 +156,182 @@ int quickSort(int arr[], int low, int high){
+       // move temporary pivot index forward
+      i++;
+      // swap current element with the element at the temporary pivot
+      swapElements(arr+i, arr+j);
+    }
+  }
+  // move the pivot element to the correct pivot position (between smaller and larger elements)
+  swapElements(arr+i+1, arr+high);
+  return (i+1);
+}
+
+// will return 0 if first param is bigger than second
+// otherwise returns 1
+int isBiggerOrSameCharset(char charset[], char charset2[]) {
+  int result;
+  // checks all 3 letters in letters array for size
+  // TODO use global variable instead of '3'
+  for (int i = 0; i < 3; i++) {
+    if (charset[i] > charset2[i]) {
+      result = 0;
+      // break here if first letter was bigger
+      break;
+    } else if (charset[i] == charset2[i]) {
+      result = 0;
+      // continue looking for the bigger letter
+    } else {
+      result = 1;
+      // break here if first letter was smaller
+      break;
+    }
+  }
+  return result;
+}
+
+int partitionWithElementsByLetters(Element arr[], int low, int high) {
+  // last Element becomes pivot
+  char *pivot = (arr+high)->letters;
+  // i is temporary pivot index - marks the index in the array, where everything with a smaller index (to the left)
+  // is already smaller than the pivot itself, but not sorted itself
+  int i = (low - 1);
+
+  // for all elements smaller than pivot move them to the left
+  for (int j = low; j < high; j++){
+    if (isBiggerOrSameCharset(pivot, (arr+j)->letters) == 0){
+       // move temporary pivot index forward
+      i++;
+      // swap current element with the element at the temporary pivot
+      swapElements(arr+i, arr+j);
+    }
+  }
+  swapElements(arr+i+1, arr+high);
+  return (i+1);
+}
+
+// executes quick sort on array of Element
+// either sorting it by numbers or letters depending on kind (either 'l' or 'n')
+int quickSortWithElements(Element arr[], int low, int high, char *kind){
+  if (low < high) {
+    int pivot;
+    // TODO is there a better way to differentiate between those two ways of sorting?
+    if (strcmp(kind, "n") == 0) {
+      // partition array and get pivot index sorted by numbers
+      pivot = partitionWithElementsByNumber(arr, low, high);
+    } else if (strcmp(kind, "l") == 0) {
+      // partition array and get pivot index sorted by letters
+      pivot = partitionWithElementsByLetters(arr, low, high);
+    }
+    // Sort the two partitions
+    quickSortWithElements(arr, low, pivot - 1, kind); // sort left side of pivot
+    quickSortWithElements(arr, pivot + 1, high, kind); // sort right side of pivot
+  }
+  return 0;
+}
+
+// searches for given integer in and array of Element of given size already sorted by numbers
+// starts searching in the middle of array and goes left or right depending on target element
+int searchForNumber(Element arr[], int target, int size) {
+  // first choose middle of array
+  int middle;
+  // if size is even, middle is offset by 1 to the right
```

Comment:
dealing with integers, dividing an uneven number by 2 should round down always automatically